topic/topic.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topic
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3msg/generated/proto/topicpb"
+)
+
+type topic struct {
+	name             string
+	numOfShards      uint32
+	consumerServices []ConsumerService
+}
+
+// NewTopic creates a new topic.
+func NewTopic() Topic { return new(topic) }
+
+// NewTopicFromValue creates a topic from a kv.Value.
+func NewTopicFromValue(v kv.Value) (Topic, error) {
+	var topic topicpb.Topic
+	if err := v.Unmarshal(&topic); err != nil {
+		return nil, err
+	}
+	return NewTopicFromProto(&topic)
+}
+
+// NewTopicFromProto creates a topic from a proto.
+func NewTopicFromProto(t *topicpb.Topic) (Topic, error) {
+	css := make([]ConsumerService, len(t.ConsumerService))
+	for i, cspb := range t.ConsumerService {
+		cs, err := NewConsumerServiceFromProto(cspb)
+		if err != nil {
+			return nil, err
+		}
+		css[i] = cs
+	}
+	return NewTopic().
+		SetName(t.Name).
+		SetNumberOfShards(t.NumOfShards).
+		SetConsumerServices(css), nil
+}
+
+func (t *topic) Name() string {
+	return t.name
+}
+
+func (t *topic) SetName(value string) Topic {
```

Comment:
NIT: this is just a naming thing. It seems like this function would set the name of the topic, however it returns a new topic with the name set, the original topic name is unchanged. I might point this out in a comment or something.